### DISCUSSION TITLE
How Horovod records events in TimeLine?

### DISCUSSION THREAD

**Discussion by orwa-te** [horovod/horovod#2787] [Q&A]
**Environment:**
1. Framework: (TensorFlow)
2. Framework version: 2.30
3. Horovod version: 0.20.3
4. MPI version: 4.0.5
5. CUDA version: N/A
6. NCCL version: N/A
7. Python version: 3.7.7
8. OS and version: Centos 7
9. GCC version: 4.8.5

**Your question:**

I am trying to debug the Horovod timeline file which was recorded on a Spark cluster using 2 worker nodes while training `one epoch` of a simple neural network with `mnist `data set that has only an input layer, one hidden dense layer, and an output layer.
After I opened the timeline JSON file in Chrome, I found the following:

![image](https://user-images.githubusercontent.com/32763039/109418044-f2e12f80-79ce-11eb-89db-c027b5487954.png)

There are 17 panes with `pid `from 1 to 17. I concluded from [this explanation](https://github.com/horovod/horovod/issues/321) that there are 17 tensors where each `pid `represent a tensor, but why? There should be about 6 tensors only in the the graph, (one Input, two weights matrices, one hidden layer, one output layer and the bias matrix).
Secondly, why are the first four panes have different events from all others?
I have not found any enough explanation that shows how things is arranged in such a way!

**Comment by tgaddair** [comment]
Hey @orwa-te, you should see 6 Allreduce panes corresponding to the 6 parameters in the graph. The rest are likely to be Broadcast calls for the model params and optimizer state.

**Comment by orwa-te** [comment]
Hello @tgaddair, thanks for the reply.
As I understand for this neural network which has **3 layers**(**1** input, **1** hidden, and **1** output), there are **4** tensors from the first node to be aggregated with the other **4** tensors in the second node: 2 for **weights** and 2 for **biases**, and this explains why **MPI_ALL_REDUCE** exists only in the first 4 panes, right?
Could you please tell me how to compute the processing time (more likely while calculating the gradients) of a batch from the timeline?
My other question is why the **interval** between the first batch and the second batch in a tensor is much bigger than intervals between other batches?

**Comment by tgaddair** [comment]
The first batch is usually slower because of initialization steps and lack of a warm cache. The timeline will not tell you the end-to-end time for the batch, only time for the part managed by Horoovd, but this info is available in the JSON file.
